CIP-optimized-uplc-serialization/README.md

```diff
@@ -0,0 +1,446 @@
+
+> (Appendix D, Section D.3.3, page 31 )
+> We define the encoder and decoder for [ constant ] types by combining 𝖾 𝗍𝗒𝗉𝖾 and 𝖽 𝗍𝗒𝗉𝖾 with 𝖤 and decoder for lists of four-bit integers (See section D.2)
+
+> (Appendix D, Section D.2.2, page 31 )
+> Suppose that we have a set 𝑋 for which we have defined an encoder 𝖤 𝑋 and a decoder 𝖣 𝑋 ; we define an 𝑋 which encodes lists of elements of 𝑋 by emitting the encodings of the elements of the list, encoder 𝖤 each preceded by a 𝟷 bit, then emitting a 𝟶 bit to mark the end of the list.
+
+from what cited above we can understand that the current constant's types serialization works exactly as the list one, in particular a list of 4 bit values.
+
+this encoding takes care also of the "empty list" case which is never the case for well formed list of types.
+
+this means that an empty list encodes as ```0``` (nil constructor + no data) and a generic (simple) type encodes as ```1 xxxx 0``` ( cons construcotr + tag + nil constructor )
+
+for lists that are non empty ( such as 7 bits lists used for unsigned integers serialization ) the constructor always carries data (since of course is non-empty) and this allows to remove the last bit.
+
+the same should be applied to constant's types so that a generic (simple) type serializes as ```0 xxxx```( nil construcotr + tag )
+
+for explanation purposes it is assumed that a constant type list is defined as:
+
+```hs
+data ConstTag = IntTag | Bs | Str | UnitTag | BoolTag | ListTag | PairTag | TyApp | DataTag
+
+data TagList = Cons ConstTag TagList | Nil
+```
+
+on which is defined an ```encode``` function that returns a ```String``` representing the encoded bits
+
+```hs
+encodeTag :: ConstTag -> String
+encodeTag tag = "xxxx"
+
+encodeTagList :: TagList -> String
+encodeTagList ( Cons tag types ) = "1" ++ encodeTag tag ++ encodeTagList types
+encodeTagList ( Nil ) = "0"
+```
+
+#### new ```type``` serialization
+
+using the same model of before the new specification would require to change the ```TagList``` definition to
+```hs
+data TagList = Cons ConstTag TagList | Nil ConstTag
+```
+
+allowing to encode as:
+```hs
+encodeTag :: ConstTag -> String
+encodeTag tag = "xxxx"
+
+encodeTagList :: TagList -> String
+encodeTagList ( Cons tag types ) = "1" ++ encodeTag tag ++ encodeTagList types
+encodeTagList ( Nil tag ) = "0" ++ encodeTag tag
+```
+
+### remove the "type application" tag for constant types
+
+as per specification constant terms are typed using tags.
+
+these tags are serialized as 4 bits and currently so 16 options ( 0 to 15 inclusive ) of wich 9 are used as follows
+
+tags from ```integer``` to ```bool``` and the ```data``` one are directly followed by the respective value encoding;
+
+tags ```list``` and ```pair``` are the only tags that do require some other tag in order to be a defined type; since the twos always require some other type in order to be valid, the type application is implcit and it should be removed.
```

Comment:
Yes, that's a bit annoying: it's like that because it reflects how the internal representations of types work.  The extra generality might permit us to do some more complex things in future, but it's not clear if we'll ever actually need that.  Simplifying this might be a good thing to do, since it complicates the encoding/decoding process as well as taking up extra room.